keps/sig-api-machinery/3157-watch-list/README.md

```diff
@@ -0,0 +1,1052 @@
+This imposes a global order of events. It is equivalent to a LIST followed by a WATCH request.
+
+Note 2: Currently, there is a timeout for LIST requests of 60s.
+That means a slow reflector might fail synchronization as well and would have to re-establish the connection.
+
+Step 2c: How bookmarks are delivered to the cacheWatcher?
+
+First of all, the primary purpose of bookmark events is to deliver the current resourceVersion to watchers, continuously even without regular events happening.
+There are two sources of resourceVersions.
+The first one is regular events that contain RVs besides objects.
+The second one is a special type of etcd event called progressNotification delivering the most up-to-date revision with the given interval only to the kube-apiserver.
+As already mentioned in 2a the watchCache is driven by the reflector.
+Every event will be eventually propagated from the watchCache to the cacher.processEvent method.
+For simplicity, we can assume that the processEvent method will simply update the resourceVersion maintained by the cacher.
+
+At regular intervals, the cacher checks expired watchers and tries to deliver a bookmark event.
+As of today, the interval is set to 1 second.
+The bookmark event contains an empty object and the current resourceVersion.
+By default, a cacheWatcher expires roughly every 1 minute.
+
+The expiry interval initially will be decreased to 1 second in this feature's code-path.
+This helps us deliver a bookmark event that is >= bookmarkAfterResourceVersion much faster.
+After that, the interval will be put back to the previous value.
+
+Note: Since we get a notification every 5 seconds from etcd and we try to deliver a bookmark every 1 second.
+It seems the maximum delay time a reflector will have to wait after receiving initial data is 6 seconds (assuming small dataset).
+It might be unlikely in practice since we might get bookmarkAfterResourceVersion even before handling initial data.
+Also sending data itself takes some time as well.
+
+Step 3: After receiving a BOOKMARK event the reflector is considered to be synchronized.
+It replaces its internal store with the collected items (syncWith) and reuses the current connection for getting further events.
+
+#### API changes
+
+Extend the optional `ResourceVersionMatch` query parameter of `ListOptions` with the following enumeration value:
+
+```
+const (
+    // ResourceVersionMatchMostRecent matches data at the most recent ResourceVersion.
+    // The returned data is consistent, that is, served from etcd via a quorum read.
+    // For watch calls, it begins with synthetic "Added" events of all resources up to the most recent ResourceVersion.
+    // It ends with a synthetic "Bookmark" event containing the most recent ResourceVersion.
+    // For list calls, it has the same semantics as leaving ResourceVersion and ResourceVersionMatch unset.
+    ResourceVersionMatchMostRecent ResourceVersionMatch = "MostRecent"
```

Comment:
Hmm - I started wondering if we really need that.

"MostRecent" is effectively the semantic of `resourceVersion=""`. We are using that semantic for List.

So instead of introducing yet another constant, I think we should just:
- set RV="" for watch calls
- handle this from watchcache (if it is enabled)

Reply:
We need to be very careful to ensure that we will be backward compatible here, but I think it's doable.